File: CustomerApp/app/src/main/java/com/example/myapplication/util/android/base/BaseAdapter.java
package com.example.myapplication.util.android.base;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import java.util.List;

/**
 * Base class for RecyclerView adapters
 * @param <VH> type of RecyclerView.ViewHolder (must extend BaseViewHolder)
 * @param <T> type of item held by the adapter
 */
public abstract class BaseAdapter<VH extends BaseViewHolder<? extends ViewBinding, T>, T> extends RecyclerView.Adapter<VH> {
    protected final List<T> items;

    protected BaseAdapter(final List<T> items) {
        this.items = items;
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        holder.bindTo(items.get(position));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    /**
     * Submit a new list of items to be added to the adapter
     * @param newItems list of new items
     * @param overwrite if the existing items must be overwritten or not. If true, items are
     *                  replaced, else they are added
     */
    public void submitList(final List<T> newItems, final boolean overwrite) {
        final int size = items.size();
        if (overwrite) {
            items.clear();
            notifyItemRangeRemoved(0, size);
            items.addAll(newItems);
            notifyItemRangeInserted(0, newItems.size());
        } else {
            items.addAll(newItems);
            notifyItemRangeInserted(size - 1, newItems.size());
        }
    }

    /**
     * Clears the data held and notifies updates to the RecyclerView
     */
    public void clear() {
        final int size = items.size();
        items.clear();
        notifyItemRangeRemoved(0, size);
    }
}

File: CustomerApp/app/src/main/java/com/neighbourschef/customer/util/android/base/BaseAdapter.kt
package com.neighbourschef.customer.util.android.base

import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding

/**
 * Base class for RecyclerView adapters
 * @param <VH> type of RecyclerView.ViewHolder (must extend BaseViewHolder)
 * @param <T> type of item held by the adapter
 */
abstract class BaseAdapter<VH: BaseViewHolder<out ViewBinding, T>, T>(
    val items: MutableList<T>,
    stableIds: Boolean
): RecyclerView.Adapter<VH>() {
    init {
        super.setHasStableIds(stableIds)
    }

    override fun onBindViewHolder(holder: VH, position: Int) = holder.bindTo(items[position])

    override fun getItemCount(): Int = items.size

    /**
     * Submit a new list of items to be added to the adapter
     * @param newItems list of new items
     * @param overwrite if the existing items must be overwritten or not. If true, items are
     * replaced, else they are added
     *
     * @see BaseAdapter.addItem
     */
    fun submitList(newItems: List<T>, overwrite: Boolean = true) {
        val size = items.size
        if (overwrite) {
            clear()
            items += newItems
            notifyItemRangeInserted(0, newItems.size)
        } else {
            items += newItems
            notifyItemRangeInserted(size - 1, newItems.size)
        }
    }

    /**
     * Add a single item to the adapter (better while adding single item instead of wrapping in a list)
     * @param item item submitted
     *
     * @see BaseAdapter.submitList
     */
    fun addItem(item: T) {
        items += item
        notifyItemInserted(items.size - 1)
    }

    /**
     * Clears the data held and notifies updates to the RecyclerView
     */
    fun clear() {
        val size = items.size
        items.clear()
        notifyItemRangeRemoved(0, size)
    }
}
